crates/types/src/traits/block_contents.rs

```diff
@@ -50,7 +46,17 @@ pub trait BlockPayload:
     /// The type of the transitions we are applying
     type Transaction: Transaction;
 
-    // type Header: BlockHeader;
+    /// Data created during block building which feeds into the block header
+    type Metadata;
+
+    /// Build a payload and associated metadata with the transactions.
+    fn build(transactions: impl IntoIterator<Item = Self::Transaction>) -> (Self, Self::Metadata);
+
+    /// Encode the payload
+    fn encode(&self) -> Vec<u8>;
+
+    /// Decode the payload
+    fn decode(encoded_transactions: &[u8]) -> Self;
```

Comment:
Is it correct for me to view `build` and `decode` as two different constructors? Maybe rename them to `from_txs` and `from_bytes`. Why does `build` return `(Self, Metadata)` whereas `decode` returns only `Self`?

Reply:
@jbearer I'm starting to think it's bad form for `build` to return `(Self, Metadata)`. I think we should view it as a constructor, so it should return only `Self`. We can add a getter to the trait to allow people to get `Metadata` if they want it. wdyt?

Reply:
That requires you to store the `Metadata` in the block payload type, which we don't necessarily want. It's redundant because the whole point of the metadata is to later put it in the header type. I don't think it's necessarily bad form to return `(Self, OtherStuff)`. The nice thing about Rust constructors is they're just functions.

I do like the rename `from_txs` and `from_bytes`. The reason for the different types is that these functions have different purposes: `build` or `from_txs` is a one-time method for bringing a block into the world, used by the builder. But once a block is in the world, we have these two different canonical representations of it: `BlockPayload` and `(Vec<u8>, Metadata)`. We can freely convert between them with `encode`/`decode`.

Reply:
Thinking about this more, maybe it's fine to require `Metadata` to be a field of the block type. This is after all just an in-memory representation. Whenever we store the payload on disk or send it over the network, we always have the option of converting to flat bytes, and then using `decode` on the other end to reconstruct it from the flat bytes and the metadata which we already have (e.g. from a header)

Reply:
Good question! I view them as constructors that do the transaction encoding/decoding in opposite ways. `build` _encodes_ the transaction, then outputs `Self` and `Metadata`, whereas `decode` _decodes_ the transactions _with_ the `Metadata` (which I missed in the input) and outputs `Self`.

I don't have a strong opinion re: whether to make `build` return `Metadata` or add a getter instead, but for `decode`, because of the opposite encoding/decoding logic, I actually think `decode` should have `Metadata` as input, rather than output (as @jbearer also suggested in [this comment](https://github.com/EspressoSystems/HotShot/pull/2045#discussion_r1389979517)).

Reply:
Ahh, I didn't see Jeb's replies until now.

Reply:
So in summary things to update are:
* Rename functions
* Add a `Metadata` getter
* `from_txns`: return `Self` only
* `from_bytes`: add `Metadata` to input

Reply:
> ... to reconstruct it from the flat bytes and the metadata which we already have (e.g. from a header)

Ah, so metadata is on-chain and known to all. What's the difference between metadata and header? Can we unify these two concepts?

Reply:
Metadata is a part of the header. Basically the header construction is allowed to take input from two different sources: `metadata`, which relates to the payload, and `parent`, which allows us to chain state updates together. These are separate until the header is built, after which you only care about the header. They are separate initially because they come from different places, metadata from the block builder and parent header from the consensus proposer.